app/uk/gov/hmrc/tai/service/UpdateNextYearsIncomeService.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2018 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.tai.service
+
+import uk.gov.hmrc.domain.Nino
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.http.logging.MdcLoggingExecutionContext._
+import uk.gov.hmrc.tai.model.TaxYear
+import uk.gov.hmrc.tai.util.constants.journeyCache.UpdateNextYearsIncomeConstants
+import uk.gov.hmrc.tai.model.cache.UpdateNextYearsIncomeCacheModel
+import scala.concurrent.Future
+
+class UpdateNextYearsIncomeService {
+
+  lazy val journeyCacheService: JourneyCacheService = JourneyCacheService(UpdateNextYearsIncomeConstants.JOURNEY_KEY)
+  lazy val employmentService: EmploymentService = EmploymentService
+  lazy val taxAccountService: TaxAccountService = TaxAccountService
+
+  def setup(employmentId: Int, nino: Nino)(implicit hc: HeaderCarrier): Future[UpdateNextYearsIncomeCacheModel] = {
+    val taxCodeIncomeFuture = taxAccountService.taxCodeIncomeForEmployment(nino, TaxYear(), employmentId)
+    val employmentFuture = employmentService.employment(nino, employmentId)
+
+    for {
+      taxCodeIncomeOption <- taxCodeIncomeFuture
+      employmentOption <- employmentFuture
+    } yield (taxCodeIncomeOption, employmentOption) match {
+      case (Some(taxCodeIncome), Some(employment)) => {
+        val model = UpdateNextYearsIncomeCacheModel(employment.name, employmentId, taxCodeIncome.amount)
+
+        journeyCacheService.cache(model.toCacheMap)
+
+        model
+      }
+      case _ => throw new RuntimeException("[UpdateNextYearsIncomeService] Could not set up next years estimated income journey")
+    }
+  }
+
+  def get(employmentId: Int, nino: Nino)(implicit hc: HeaderCarrier): Future[UpdateNextYearsIncomeCacheModel] = {
+    journeyCacheService.currentCache flatMap {
+      case cache: Map[String, String] if cache.isEmpty => setup(employmentId, nino)
+      case cache: Map[String, String] => {
+        if (cache.contains(UpdateNextYearsIncomeConstants.NEW_AMOUNT)) {
```

Comment:
This code could potentially be improved if a solution is thought of for the multiple employments issue.